docs/USAGE-GUIDE.md

```diff
@@ -1929,6 +1929,95 @@ s2n-tls will consider all allocated memory available for use, and will attempt t
+
+The maximum early data allowed and cipher suite can be set with **s2n_psk_configure_early_data**. If the connection will negotiate an application protocol then the expected application protocol can be set with **s2n_psk_set_application_protocol**. 
+
+### Sending early data
+
+To send early data, your application should call **s2n_send_early_data** before it calls **s2n_negotiate**. Like other IO functions, **s2n_send_early_data** can potentially fail repeatedly with a blocking error before it eventually succeeds: see [I/O Functions](#io-functions) for more information.
+
+**s2n_connection_get_remaining_early_data_size** can be called to check how much more early data the client is allowed to send. If **s2n_send_early_data** exceeds the allowed maximum, s2n-tls returns a usage error.
+
+An application can stop calling **s2n_send_early_data** at any time, even if the function has not returned success yet. If **s2n_send_early_data** does return success, the connection is ready to complete the the handshake and begin sending normal data. However, **s2n_send_early_data** can continue to be called to send more early data if desired.
+
+Once a client finishes sending early data, you call **s2n_negotiate** to complete the handshake just as you would for a handshake that did not include early data.
+
+For example:
+```
+uint8_t early_data[] = "early data to send";
+ssize_t total_data_sent = 0, len = sizeof(early_data);
+while (total_data_sent < len) {
+    ssize_t data_sent = 0;
+    int r = s2n_send_early_data(client_conn, early_data + total_data_sent,
+            len - total_data_sent, &data_sent, &blocked);
+    total_data_sent += data_sent;
+    if (r == S2N_SUCCESS) {
+        break;
+    } else if (s2n_error_get_type(s2n_errno) != S2N_ERR_T_BLOCKED) {
+        exit(1);
+    }
+}
+while (s2n_negotiate(client_conn, &blocked) != S2N_SUCCESS) {
+    if (s2n_error_get_type(s2n_errno) != S2N_ERR_T_BLOCKED) {
+        exit(1);
+    }
+}
+```
+
+### Receiving early data
+
+To receive early data, your application should call **s2n_recv_early_data** before it calls **s2n_negotiate**. Like other S2N IO functions, **s2n_recv_early_data** can potentially fail repeatedly with a blocking error before it eventually succeeds: see [I/O Functions](#io-functions) for more information.
+
+Once **s2n_recv_early_data** has been called once, it must be called until it returns success. If an application stops calling **s2n_recv_early_data** early, some early data may be left unread and cause later calls to **s2n_negotiate** to fail. Calling **s2n_recv_early_data** again after it returns success is possible but does nothing.
```

Comment:
`some early data may be left unread and cause later calls to **s2n_negotiate** to fail`

will `s2n_negotiate` eventually succeed or will the error be some non-blocking status which should indicate the caller abort?